src/interpolate.f90

```diff
@@ -0,0 +1,177 @@
+        endif
+    end subroutine construct_splines_1d
+
+
+    subroutine destroy_splines_1d(spl)
+        type(SplineData1D), intent(inout) :: spl
+
+        deallocate(spl%coeff)
+    end subroutine destroy_splines_1d
+
+
+    subroutine evaluate_splines_1d(x, spl, y)
+        real(8), intent(in) :: x
+        type(SplineData1D), intent(in) :: spl
+        real(8), intent(out) :: y
+
+        real(8) :: x_norm, x_local, local_coeff(0:MAX_ORDER)
+        integer :: interval_index, k_power
+
+        x_norm = x/spl%h_step
+        interval_index = max(0, min(spl%num_points-1, int(x_norm)))
+        x_local = (x_norm - dble(interval_index))*spl%h_step  ! Distance to grid point
+
+        local_coeff(0:spl%order) = spl%coeff(:, interval_index+1)
+
+        ! Start with largest power and then multiply recursively
+        y = local_coeff(spl%order)
+        do k_power = spl%order, 0, -1
+            y = local_coeff(k_power) + x_local*y
+        enddo
+    end subroutine evaluate_splines_1d
+
+
+    subroutine construct_splines_2d(x1, x2, y, order, periodic, spl)
+        real(8), intent(in) :: x1(:), x2(:), y(:,:)
+        integer, intent(in) :: order(2)
+        logical, intent(in) :: periodic(2)
+
+        type(SplineData2D), intent(out) :: spl
+
+        real(8), dimension(:,:), allocatable  :: splcoe
+
+        integer :: i1, i2  ! Loop indices for points (1 ... num_points)
+        integer :: k1, k2  ! Loop indices for polynomial order (0 ... order)
+
+        spl%order = order
+        spl%periodic = periodic
+        spl%num_points(1) = size(x1)
+        spl%num_points(2) = size(x2)
+        spl%h_step(1) = x1(2) - x1(1)
+        spl%h_step(2) = x2(2) - x2(1)
+
+        allocate(spl%coeff(0:order(1), 0:order(2), &
+                 spl%num_points(1), spl%num_points(2)))
+
+        ! Spline over x2
+        allocate(splcoe(0:spl%order(2), spl%num_points(2)))
+        do i1=1,spl%num_points(1)
+            splcoe(0,:) = y(i1, :)
+            if (spl%periodic(2)) then
+                call spl_per(spl%order(2), spl%num_points(2), spl%h_step(2), splcoe)
+            else
+                call spl_reg(spl%order(2), spl%num_points(2), spl%h_step(2), splcoe)
+            endif
+            do k2 = 0, spl%order(2)
+                spl%coeff(1, k2, i1, :) = splcoe(k2, :)
+            enddo
+        enddo
+        deallocate(splcoe)
+
+        ! Spline over x1
+        allocate(splcoe(0:spl%order(1), spl%num_points(1)))
+        do i2=1,spl%num_points(2)
+            do k2=0,spl%order(2)
+                splcoe(0,:) = spl%coeff(1, k2, :, i2)
+                if(spl%periodic(1)) then
+                    call spl_per(spl%order(1), spl%num_points(1), spl%h_step(1), splcoe)
+                else
+                    call spl_reg(spl%order(1), spl%num_points(1), spl%h_step(1), splcoe)
+                endif
+                do k1 = 0, spl%order(1)
+                    spl%coeff(k1, k2, :, i2) = splcoe(k1, :)
```

Comment:
Does `spl%coeff(:, k2, :, i2) = splcoe` not work or do you want it to be more explicit?